Add gear icon to SAM legend etc

moontrip opened the pull request:
This will address https://github.com/VEuPathDB/web-monorepo/issues/745. Below is a list of works involved in this PR.

- [x] not to open map nav panel initially
- [x] making gear icon component
- [x] adding gear icon to all SAM legends (donut, barplot, and bubble markers)
- [x] properly arranging gear icon
- [x] mouse-over tooltip on the icon
- [x] functionality to open the map nav panel when pressing the icon
- [x] displaying snackbar when the nav panel is already open

Note: 
1) I thought that the icon was better placed at title area, not the bottom right. 
2) I used "warning" style for snackbar, but not quite sure if it is better than other styles: there are four categories such as "info", "success", "warning", and "error".  

A screenshot is attached:
![SAM legend gear icon01](https://github.com/VEuPathDB/web-monorepo/assets/12802305/92c4bc77-9115-4898-ab4e-9c648c406c82)

bobular reviewed (commented):
Hi. This is a great start. I have a few suggestions that will hopefully maintain maintainability!

`DraggablePanel` currently has higher level app-specific verbiage in it ("Marker configuration panel is already open") and an app-specific prop `isSidePanelExpanded`. It would be best if the `onClick` was just `onClick={openPanel}` and the `openPanel` callback defined in `MapAnalysis` could do all the snackbar logic. Then there's no need to pass `isSidePanelExpanded`.

I think it's OK for all `DraggablePanel`s to have an optional `SettingsButton` - I would rename its `openPanel` prop to `onSettingsButtonClick` and probably add a `settingsButtonTooltip` prop. But then we still have to pass two props around everywhere.

Would it be easier just to have a `headerButtons` prop to `DraggablePanel`? We would create the button `openMarkerConfigButton` in `MapAnalysis` which already has tooltip and onClick handler it needs. We would then just pass `openMarkerConfigButton` as a prop into the XXXMarkerMapType plugins.

We also need to set `activeSideMenuId` in the onClick handler so that the marker config panel actually opens. At the moment it will open the filter panel if that was the last thing opened.

moontrip commented:
Thank you for your review and detailed comments, @bobular 👍 I will check it out carefully next week. Have a great weekend!

moontrip commented:
I have addressed most of your feedbacks and suggestions. Please find my comments in the following that start from **DK:**

> `DraggablePanel` currently has higher level app-specific verbiage in it ("Marker configuration panel is already open") and an app-specific prop `isSidePanelExpanded`. It would be best if the `onClick` was just `onClick={openPanel}` and the `openPanel` callback defined in `MapAnalysis` could do all the snackbar logic. Then there's no need to pass `isSidePanelExpanded`.

**DK:** That makes sense. I have complied with.

> I think it's OK for all `DraggablePanel`s to have an optional `SettingsButton` - I would rename its `openPanel` prop to `onSettingsButtonClick` and probably add a `settingsButtonTooltip` prop. But then we still have to pass two props around everywhere.
> 

**DK:** This and below one are conflicted each other. Thus, I employed the next one

> Would it be easier just to have a `headerButtons` prop to `DraggablePanel`? We would create the button `openMarkerConfigButton` in `MapAnalysis` which already has tooltip and onClick handler it needs. We would then just pass `openMarkerConfigButton` as a prop into the XXXMarkerMapType plugins.
> 

**DK:** Made a headerButton component at MapAnalysis and passed it to DraggablePanel

> We also need to set `activeSideMenuId` in the onClick handler so that the marker config panel actually opens. At the moment it will open the filter panel if that was the last thing opened.

**DK:** I could manage to figure this out and now corresponding marker config panel is open regardless of the previous position.

bobular commented on packages/libs/eda/src/lib/map/analysis/MapAnalysis.tsx:
```diff
@@ -765,6 +765,56 @@ function MapAnalysisImpl(props: ImplProps) {
+          variant: 'info',
+        })
+      : openPanelBehavior();
+  }, [
```
There are a bunch of "indirect" dependencies listed here.

I would probably be better to either create `openPanelBehavior` with useCallBack with the dependency [`appState.activeMarkerConfigurationType`] (the useState-derived setters are stable and not needed), and then add `openPanelBehavior` as a dependency for the `openPanel` useCallback

Or... Just paste the code from `openPanelBehavior` into `openPanel`, double-checking the necessity of the dependencies as mentioned above. This would be fine because I don't see `openPanelBehavior` used elsewhere.

moontrip replied:
Hi @bobular! Thank you for your detailed reviews 👍 Actually I noticed that indirect dependencies this morning and as you suggested I think we can just use single openPanel useCallback. Updated accordingly.

Also, I just addressed the first item, "There's just one behavioural problem, where if the side panel is already expanded and on, say, the Filter panel, then it won't switch to the marker config panel when the gear icon is clicked." It is now switched to marker configuration with snackbar behavior. I think that the other UI/UXs tuffs can be handled in other ticket #785 as you suggested.

bobular replied:
Hi - that's still not quite the correct behaviour, IMO.

I was thinking more like this

```
  const openPanel = useCallback(() => {
    const nextSideMenuId = 'single-variable-' +
           (appState.activeMarkerConfigurationType === 'barplot'
          ? 'bar'
          : appState.activeMarkerConfigurationType);
    
    if (appState.isSidePanelExpanded && activeSideMenuId === nextSideMenuId)
      enqueueSnackbar('Marker configuration panel is already open', {
        variant: 'info',
      });
    if (!appState.isSidePanelExpanded)
      setIsSidePanelExpanded(true);
    if (activeSideMenuId !== nextSideMenuId)
      setActiveSideMenuId(nextSideMenuId);
  }, [
    appState.isSidePanelExpanded,
    enqueueSnackbar,
    activeSideMenuId,
    setIsSidePanelExpanded,
    appState.activeMarkerConfigurationType,
  ]);
```

bobular reviewed (commented):
This looks really nice now!

There's just one behavioural problem, where if the side panel is already expanded and on, say, the Filter panel, then it won't switch to the marker config panel when the gear icon is clicked.

A smaller problem is that if the marker config panel was last open in "supporting plots" sub-panel mode, then the marker config panel is opened in this mode, not the "markers" sub-panel.

A more sophisticated/flexible approach to UI state saving is probably needed long-term (see #785). Instead of `appState.activeMarkerConfigurationType` and non-persisted `activeSideMenuId` we probably need something like `appState.sidePanelNavigationState` that stores the full menu state, in a breadcrumb array or similar.  appState.isSidePanelExpanded could continue to be toggle.

Short-term, perhaps we can make it work with a function `isMarkerConfigOpen(appState.isSidePanelExpanded, appState.activeMarkerConfigurationType, activeSideMenuId)` ?  (not sure if the second arg is needed)

We can leave the problem of the "supporting plots" sub-tab to a wider-reaching ticket, such as #785

bobular commented on packages/libs/eda/src/lib/map/analysis/mapTypes/plugins/BubbleMarkerMapType.tsx:
```diff
@@ -786,5 +786,5 @@ function useMarkerData(props: DataProps) {
       },
     };
-  }, [rawMarkersResult, legendDataResult.data, overlayConfig]);
+  }, [rawMarkersResult, legendDataResult.data, overlayConfig, boundsZoomLevel]);
```
Good catch!

moontrip replied:
Thanks @bobular! 👍 So, you don't want to show snackbar if the side panel is already open and does not point out marker configuration panel (Donuts, Bar plots, or Bubbles), right? This is updated~

moontrip replied:
Thank you for your confirmation, @bobular  👍 I will merge this PR, finally 🎉

bobular reviewed (approved):
Almost what I was imagining!

I'm going to pre-emptively approve. Could you double check the logic I proposed? It works for me. Thanks!
